Ditch js/* and css/* views prefixes

ewinslow opened the pull request:
In the spirit of the-smallest-change-that-gets-it-done approach, here is round 1 of view aliasing that gets rid of the js/\* and css/\* view prefixes in a BC way.

All views under these directories are implicitly "hoisted" up one level.
They are also given the appropriate extension:
- js/view => view.js
- css/view => view.css

The main benefit this brings is being able to co-locate related assets.
So a template (view.php) can have its CSS/JS dependencies right next to it
(view.css, view.js).

BREAKING CHANGE:
Great care has been taken to make this change as backwards-compatible as possible,
so you should not need to update any view references right away. However, you are
certainly encouraged to move your JS and CSS views to their new, canonical
locations.

Certain uses of the `view,$view_name` and `view_vars,$view_name` hooks will not work.
See docs/guides/upgrading.rst for more info.

Refs #8381
Fixes #8382
- [x] Remove warning in canonicalizeViewName()
- [x] Update developers_wrap_views()
- [x] Limit docs to explaining how existing views will be affected
- [x] Make sure that views like "js/jquery.form.php" get aliased correctly
  - [x] Add a test for this
- [x] Check what's going on with 'ckeditor/init' view.

ewinslow committed:
feature(assets): Get rid of js/ and css/ directories

All views under these directories are implicitly "hoisted" up one level.
They are also given the appropriate extension:

 * js/view => view.js
 * css/view => view.css

The main benefit this brings is being able to co-locate related assets.
So a template (view.php) can have its CSS/JS dependencies right next to it
(view.css, view.js). Compare to the current approach which is like so:

 * Template: page/components/module.php
 * CSS: css/elements/modules.php
 * JS: None in this case, but how would you know with any confidence?

BREAKING CHANGE:

Great care has been taken to make this change as backwards-compatible as possible,
so you should not need to update any view references right away. However, you are
certainly encouraged to move your JS and CSS views to their new, canonical
locations.

Certain uses of the `view,$view_name` and `view_vars,$view_name` hooks will not work.
See the docs on "View aliases" for more info.

Refs #8381
Fixes #8382

ewinslow commented:
This doesn't exhibit any obvious problems on my local instance. Woohoo!

ewinslow commented:
Round two of view aliasing will be to support aliasing arbitrary views to other views. E.g.

```php
$aliases = [
  'elements/buttons.css' => 'elgg/buttons.css'
  // etc...
];
```

The aliases would be hard-coded in core at first, then phase 3 would be to open up aliasing to plugins if we feel the benefits would be worth the complexity...

ewinslow commented:
@Elgg/owners @Elgg/contributors Would love some quick feedback on this.

jdalsem commented on engine/classes/Elgg/ViewsService.php:
```diff
@@ -54,11 +54,46 @@ public function __construct(\Elgg\PluginHooksService $hooks, \Elgg\Logger $logge
+		}
+		
+		if ($alias !== $canonical) {
+			$message = "The deprecated view $alias was moved to $canonical.\n";
```
probably should use a elgg_echo

ewinslow replied:
Most deprecation errors don't use elgg echo, iirc.

mrclay replied:
Is the notice really necessary? Pretty much all non-core plugins are going to generate lots of them and we'll nudge them into making their plugin non-BC with 1.12.

mrclay replied:
Do we have a strong motivation for dropping support in 3.0?

ewinslow replied:
We can retract the warning. I just figured in general aliasing ==
complicated so using canonical name is "better".

On Mon, Jun 22, 2015, 4:13 AM Steve Clay <notifications@github.com> wrote:

> In engine/classes/Elgg/ViewsService.php
> <https://github.com/Elgg/Elgg/pull/8526#discussion_r32923872>:
>
> > +		$canonical = $alias;
> > +
> > +		if (strpos($canonical, "js/") === 0) {
> > +			$canonical = substr($canonical, 3);
> > +			if (!pathinfo($canonical, PATHINFO_EXTENSION)) {
> > +				$canonical .= ".js";
> > +			}
> > +		} else if (strpos($canonical, "css/") === 0) {
> > +			$canonical = substr($canonical, 4);
> > +			if (!pathinfo($canonical, PATHINFO_EXTENSION)) {
> > +				$canonical .= ".css";
> > +			}
> > +		}
> > +		
> > +		if ($alias !== $canonical) {
> > +			$message = "The deprecated view $alias was moved to $canonical.\n";
>
> Do we have a strong motivation for dropping support in 3.0?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8526/files#r32923872>.
>

mrclay replied:
I added a todo item for this

jdalsem commented:
Code looks ok. Not tested yet

mrclay commented on docs/guides/upgrading.rst:
```diff
@@ -12,6 +12,42 @@ See the administrator guides for :doc:`how to upgrade a live site </admin/upgrad
+
+In order to make this change, we had to re-configure view location storage such
+that full paths are stored, not just paths to view directories. 
+So ``elgg_get_view_location`` and ``elgg_set_view_location`` return and expect full
```
I think better to no-op these and spit out an error, and come up with new functions.

mrclay replied:
elgg_set_view_location() would be easy to keep working, and elgg_get_view_location() could still work in many cases, though best to get rid of it.

mrclay commented on engine/classes/Elgg/Application/CacheHandler.php:
```diff
@@ -83,7 +83,7 @@ public function handleRequest($path) {
 			exit;
 		}
 
-		$filename = $config->getVolatile('dataroot') . 'views_simplecache/' . md5("$viewtype|$view");
```
This is needed to limit filename length

ewinslow replied:
Why does length need to be limited?

mrclay replied:
Windows

ewinslow replied:
I don't follow, how is it that we can have long views in the views directory but then when we cache them they have to be shortened?

mrclay replied:
Good point. This was probably a vestige of my experience caching for Minify, as Minify tries hard to name cache files so that you could recognize the files in them. Weird I know.

mrclay commented on engine/classes/Elgg/HooksRegistrationService.php:
```diff
@@ -56,6 +56,10 @@ public function registerHandler($name, $type, $callback, $priority = 500) {
 		}
+		
+		if (($name == 'view' || $name == 'view_vars') && $type != 'all') {
+			$type = _elgg_services()->views->canonicalizeViewName($type);
```
I think this will break lots of handlers. Many I've written expect particular view names passed in.

mrclay replied:
This is my biggest concern with this. I think it would create an especially silent class of bugs, especially if you register for [views, all].

ewinslow replied:
So you register for views/all and then check for particular view names?

mrclay commented on engine/lib/views.php:
```diff
@@ -1622,19 +1622,23 @@ function _elgg_views_send_header_x_frame_options() {
  */
 function elgg_views_boot() {
+	if (!elgg_get_config('system_cache_loaded')) {
+		_elgg_services()->views->registerPluginViews(realpath(__DIR__ . '/../../'));
```
:+1:

ewinslow replied:
Core is just the first plugin loaded :)

ewinslow commented:
So the main thing here is concern over hook handlers for the view and view_vars hooks.

I made it so if you register or a particular view, that should work as expected. However if you register for "all" and then check the view structure or name manually, you're probably going to have a bad time.

Open to suggestions about how to handle that case...

mrclay commented:
Well, if you register for [view, js/foo] your handler will get called, but `$type` won't be what you're expecting. Not always is it ignored.

mrclay commented:
I'm not necessarily in favor of blocking this, but all the implications need to be documented.

mrclay commented:
Please let's try to finish up #8513 (it's very close) before we consider this.

ewinslow commented:
Sounds fine

On Fri, Jun 19, 2015, 3:22 PM Steve Clay <notifications@github.com> wrote:

> Please let's try to finish up #8513
> <https://github.com/Elgg/Elgg/pull/8513> (it's very close) before we
> consider this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8526#issuecomment-113660818>.
>

ewinslow commented:
Updated docs to mention some gotchas

mrclay commented:
Could we use a real example like css/elgg  instead of foo and bar? Or some other real view that is likely to be used in view hooks?

ewinslow commented:
Good call, will switch to css/elgg...

On Sat, Jun 20, 2015, 7:34 PM Steve Clay <notifications@github.com> wrote:

> Could we use a real example like css/elgg instead of foo and bar? Or some
> other real view that is likely to be used in view hooks?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8526#issuecomment-113853985>.
>

ewinslow commented:
Made a few more updates.

@mrclay At what point would you feel comfortable pulling this in?

mrclay commented:
I think we need more feedback from plugin devs. This yields a small DX improvement but potentially break plugins. I'd be more comfortable if 1.12 could deprecate listeners like [view, css/elgg] and maybe mod/developers could warn on listening for [view, all].

mrclay commented:
Make the case for this on Community maybe...

ewinslow commented:
Good call. Will do.

ewinslow commented:
https://community.elgg.org/discussion/view/2219149/elgg-20-remove-js-and-css-view-prefixes-in-a-mostly-bc-way

On Sun, Jun 21, 2015 at 9:32 AM Steve Clay <notifications@github.com> wrote:

> Make the case for this on Community maybe...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8526#issuecomment-113923436>.
>

mrclay commented:
> elgg_get_view_location and elgg_set_view_location return and expect full paths respectively

No longer true.

ewinslow commented:
Hmmm... Thought I had removed those.

On Mon, Jun 22, 2015, 5:06 PM Steve Clay <notifications@github.com> wrote:

> elgg_get_view_location and elgg_set_view_location return and expect full
> paths respectively
>
> No longer true.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8526#issuecomment-114309274>.
>

ewinslow commented:
Updated pr description and commit messages

On Mon, Jun 22, 2015, 5:14 PM Evan Winslow <evan@elgg.org> wrote:

> Hmmm... Thought I had removed those.
>
> On Mon, Jun 22, 2015, 5:06 PM Steve Clay <notifications@github.com> wrote:
>
>> elgg_get_view_location and elgg_set_view_location return and expect full
>> paths respectively
>>
>> No longer true.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/Elgg/Elgg/pull/8526#issuecomment-114309274>.
>>
>

ewinslow commented:
@jdalsem any chance you'd be able to treat this on some of your plugins?

mrclay commented:
Does this update `developers_wrap_views` as needed?

ewinslow commented:
Can you add a to-do so I don't forget to double check?

On Tue, Jun 23, 2015, 11:02 AM Steve Clay <notifications@github.com> wrote:

> Does this update developers_wrap_views as needed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8526#issuecomment-114591912>.
>

mrclay commented on docs/guides/views.rst:
```diff
@@ -358,6 +360,73 @@ Here we'll eliminate breadcrumbs that don't have at least one link.
 	    // returning nothing means "don't alter the returnvalue"
 	}
 
+View aliases
```
I think these docs should be limited to explaining how existing views will be affected. They're a bit aspirational for a feature it's not clear we're going to add.

ewinslow replied:
Will do

ewinslow commented on engine/classes/Elgg/ViewsService.php:
```diff
@@ -89,37 +89,73 @@ public function __construct(\Elgg\PluginHooksService $hooks, \Elgg\Logger $logge
+
+		if (strpos($canonical, "js/") === 0) {
+			$canonical = substr($canonical, 3);
+			if (!pathinfo($canonical, PATHINFO_EXTENSION)) {
```
So, I think this will catch views like jquery.ui or jquery.form and think they have an extension like "ui" or "form", and therefore fail to add ".js". We should check the extensions against a whitelist to make sure they are recognized, then move them.

mrclay replied:
Sounds good

mrclay commented:
This gets a :+1: for me. I think a special guide for upgrading hook handlers would be really cool. Searching the codebase for `"view"` and `'view'` worked pretty well for finding handler registration.

mrclay commented on js/lib/configuration.js:
```diff
@@ -15,6 +15,7 @@ elgg.get_site_url = function() {
  * @param {String} view    The full view name
  * @param {String} subview If the first arg is "css" or "js", the rest of the view name
  * @return {String} The site URL.
+ * @deprecated Use "require.toUrl(view)" inside your AMD modules instead.
```
Are we sure toUrl() is sufficient? `foo/bar` might be a valid view for ajax HTML content, but I wonder if toUrl() will add `.js`...

ewinslow replied:
Removing...

ewinslow commented:
@Elgg/owners @Elgg/contributors anyone else care to give this a once-over before I pull the trigger?

mrclay commented on engine/tests/phpunit/Elgg/Cache/SimpleCacheTest.php:
```diff
@@ -29,6 +29,10 @@ public function testGetUrlHandlesTwoArgumentsWhereSecondArgHasRedundantPrefix()
 		$url = $simpleCache->getUrl('js', 'js/view.js');
 		
-		$this->assertTrue(preg_match("#default/js/view.js$#", $url));
+		$this->assertTrue(preg_match("#default/view.js$#", $url));
```
`\.`

mrclay commented:
Do we have a test for the view "js/jquery.form.php"?

ewinslow commented:
Added to-do, thanks again for keeping me honest.

juho-jaakkola commented:
ckeditor/init.php seems to be causing js errors

mrclay commented:
Just a heads up, I can tell this will need to be carefully rebased after #8553

juho-jaakkola commented:
I will test and review some more once the javascript error has been fixed.

ewinslow commented:
Turns out it was actually a problem with the "languages" file, which was getting 2 anonymous "define" calls in the same file somehow. I really wish requirejs had a more helpful error message there, like telling you which script had the define mismatch, because that was annoying to debug.

ewinslow commented:
In any case, problem should be fixed. Eager to pull this in!

mrclay commented:
Tested fine for me. @juho-jaakkola if this works for you, feel free to merge.

ewinslow commented:
@juho-jaakkola Please do test again if you feel so inclined. This is blocking/conflicting with other PRs and is otherwise approved, so I'm going to pull this in. I swear I will be more patient after the 2.0 feature freeze date!
